include/circt/Dialect/Emit/EmitOps.td

```diff
@@ -0,0 +1,114 @@
+//===- EmitOps.td - `emit` dialect ops ---------------------*- tablegen -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This describes the MLIR ops for `emit`.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef CIRCT_DIALECT_EMIT_EMITOPS_TD
+#define CIRCT_DIALECT_EMIT_EMITOPS_TD
+
+include "circt/Dialect/HW/HWTypes.td"
+include "circt/Dialect/Emit/EmitDialect.td"
+include "mlir/Interfaces/SideEffectInterfaces.td"
+
+class EmitOp<string mnemonic, list<Trait> traits = []> :
+    Op<EmitDialect, mnemonic, traits>;
+
+def FileOp : EmitOp<"file", [
+    Symbol,
+    SingleBlock,
+    NoTerminator,
+    NoRegionArguments,
+    IsolatedFromAbove
+]> {
+  let summary = "Represents the contents of an emitted file";
+
+  let description = [{
+    This operation groups a set of nested operations to be emitted to a file.
+  }];
+
+  let regions = (region SizedRegion<1>:$body);
+  let arguments = (ins
+    StrAttr:$file_name,
+    OptionalAttr<SymbolNameAttr>:$sym_name
+  );
+  let results = (outs);
+
+  let assemblyFormat = "$file_name (`sym` $sym_name^)? $body attr-dict";
+
+  let skipDefaultBuilders = 1;
+  let builders = [
+    // Creates a file with a callback. While the callback is executed, the
+    // insertion point of the builder is moved inside the body of the file op.
+    OpBuilder<(ins "StringRef":$symName, "StringRef":$fileName,
+                    CArg<"llvm::function_ref<void()>">:$bodyCtor)>,
+    OpBuilder<(ins "StringRef":$fileName,
+                    CArg<"llvm::function_ref<void()>">:$bodyCtor)>,
+  ];
+
+  let extraClassDeclaration = [{
+    // SymbolOpInterface
+    static bool isOptionalSymbol() { return true; }
+
+    // Utilities
+    Block *getBodyBlock() { return &getBodyRegion().front(); }
+  }];
+}
+
+def VerbatimOp : EmitOp<"verbatim"> {
+  let summary = "Verbatim opaque text emitted inline.";
+  let description = [{
+    This operation produces opaque text inline in the file.
+
+    emit.verbatim allows symbol reference substitutions with {{0}} syntax.
+  }];
+  let arguments = (ins
+    StrAttr:$format_string,
+    OptionalAttr<NameRefArrayAttr>:$symbols
```

Comment:
My worry is that what it means to substitute a symbol in a verbatim depends on the two dialects involved.  Does emit.verbatim know about sv symbols and ops and define all those substitutions, or do ops specify how to substitute themselves?  For both of these, what if the correct substitution depends on the dialect of the verbatim and the dialect of the op with the symbol?

Which is a long ways of saying I see how `sv.verbatim` knows what to do for sv.module symbols.  If the verbatim is for sv code, why not leave the verbatim there?  There are couplings between ops.